### PR TITLE
Chat filter

### PR DESCRIPTION
**Pull request by loveridge**
This chat filter resolves a lot of the objections I saw with the last one that was proposed. It aims to inform players that some language is not acceptable in BAR, and to reduce moderator burden. The goal is not to prevent all chat abuse, as that is not possible. I think DeviousNull has written a much better reflection on chat filters than I am capable of, and described what this implementation hopes to achieve, in this comment: https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3539#issuecomment-2277226295

A list of the most problematic words from the moderation team would be nice to fill out the word list.

### CODE REVIEW

File: luaui/Widgets/gui_chat.lua [hunk 1]
```diff
@@ -453,6 +453,21 @@ local function getAIName(teamID)
 	return Spring.I18N('ui.playersList.aiName', { name = name })
 end
 
+local lastMessage
+local badWords = {
+	"^retard[s]?$",
```
Comment (sprunk) [inline luaui/Widgets/gui_chat.lua:458]: nigger

**Comment by WatchTheFort** [comment]
If a message is not allowed, then it is not allowed. It doesn't become okay the second time it is sent.

**Comment by WatchTheFort** [comment]
The problem is that it also allows true positives to bypass the filter.

**Comment by loveridge** [comment]
> The problem is that it also allows true positives to bypass the filter.

Sure, and then that person can be reported to have a moderator take action. When a player can just bypass the filter by adding a period or something to the word, then there is no point in preventing the player from sending the message if they really want to. When someone has intentionally bypassed the filter, then they have ignored the warning and can be given harsher consequences.

**Comment by sprunk** [comment]
I'm sure a bunch of edgy 16 year olds will be able to break any filter you think is "very good" within like 3 minutes tops.

**Comment by Beherith** [comment]
I like that its being blocked from the sending side. Clever use of repeat. Repeats could be auto-flagged.

**Comment by loveridge** [comment]
I've added an initial word list and ROT13'd it as requested.

**Comment by S3KCentrifugal** [comment]
Still worth putting the widget in even if people get around it. If you want to reduce moderator burden I would suggest batching messages up to a server and running jobs against all the logs. Ideally you would have a way in real time. However post processing logs will still take care of bans without moderator involvement.

This also sounds like a good use case for language models for more complex analysis.

**Comment by WatchTheFort** [comment]
Do we plan on the chat filter supporting more than just English?

File: luaui/Widgets/gui_chat.lua [hunk 2]
```diff
@@ -1888,7 +1901,14 @@ function widget:KeyPress(key)
 					if ssub(inputText, 1, 1) == '/' then
 						Spring.SendCommands(ssub(inputText, 2))
 					else
-						Spring.SendCommands("say "..inputMode..inputText)
+						local badWord = findBadWords(inputText)
+						if badWord ~= nil and inputText ~= lastMessage then
+							addChat(Spring.GetGameFrame(), LineTypes.System, "\255\255\000\000Moderation:",
+							'Words such as "' .. badWord .. '" can be against the Code of Conduct if they are being used to abuse other players. Please take a moment to reevaluate your intentions. If your message was blocked in error, then you can send it again to bypass the filter.', true)
```
Comment (WatchTheFort) [inline luaui/Widgets/gui_chat.lua:1907]: This message needs to support I18N. We should not be hardcoding any display text.

File: luaui/configs/badwords.lua
Comment (WatchTheFort) [inline luaui/configs/badwords.lua:1]: Can this file get a comment on why it's implemented like this

**Review by WatchTheFort — CHANGES_REQUESTED** [review]

**Comment by loveridge** [comment]
I haven't had a chance to resolve the comments yet

**Comment by WatchTheFort** [comment]
@loveridge Will you have an opportunity to work on this, or should it be reassigned?

**Comment by RogShotz** [comment]
I'd like to put my two cents in that players who would like to have a filter on can have a filter on, and players who do not do not. People can be just as mean without using a naughty word, and if anything, I've found it to increase toxicity in almost every game I've ever played. I would highly recommend just having it on by default to censor something to asterisks' (*****) instead. Simply not sending it and saying "don't say bad things" will reduce the moderation burden by 0%.

Another note, that censoring is something BAR should have regardless as an option since i'm sure kids play this etc. It's also good for content creators.

For slurs just automute them for a day or auto make a report. Censorship is a slippery slope, with asterisks no one loses anything.

'One man's vulgarity is another's lyric'.

**Comment by RogShotz** [comment]
Can we perhaps have a list of the words as well in plain English so the community can be more informed about what words are considered bad or not btw?

Also, I think this is a good censorship method. I'd recommend that there also be a profanity filter option that asterisks any bad words for kids, content creators, etc. Would help to keep the game age appropriate for everyone. Good patch, might pick this up next.

**Comment by WatchTheFort** [comment]
> Can we perhaps have a list of the words as well in plain English so the community can be more informed about what words are considered bad or not btw?

No, the whole point of the ROT13 is to _not_ have those words in our codebase. If you're thinking of using a word and are not sure if it's allowed, then don't use it.

**Comment by obj-obj** [comment]
> > Can we perhaps have a list of the words as well in plain English so the community can be more informed about what words are considered bad or not btw?
> 
> No, the whole point of the ROT13 is to _not_ have those words in our codebase. If you're thinking of using a word and are not sure if it's allowed, then don't use it.

Was about to ask why the alphabet was being rotated, this makes a lot of sense
